paperqa/rate_limiter.py

```diff
@@ -0,0 +1,382 @@
+
+        Note the key is created with RateLimitItem.key_for(*identifiers),
+        the first key is the namespace, then the next two will be our identifiers.
+
+        """
+        namespace, primary_key = key.split("/")[1:3]
+        rate_limit, new_namespace = self.parse_rate_limits_and_namespace(
+            namespace, primary_key
+        )
+        return (
+            rate_limit,
+            (new_namespace, primary_key),
+        )
+
+    async def get_rate_limit_keys(
+        self, cursor_scan_count: int = 100
+    ) -> list[tuple[RateLimitItem, tuple[str, str | MatchAllInputs]]]:
+        """Returns a list of current RateLimitItems with tuples of namespace and primary key."""
+        host, port = os.environ.get("REDIS_URL", ":").split(":", maxsplit=2)
+
+        if not (host and port):
+            raise ValueError(f'Invalid REDIS_URL: {os.environ.get("REDIS_URL")}.')
+
+        client = Redis(host=host, port=int(port))
+
+        cursor = b"0"
+        matching_keys: list[bytes] = []
+
+        while cursor:
+            cursor, keys = await client.scan(
+                int(cursor), match=f"{self.storage.PREFIX}*", count=cursor_scan_count
+            )
+            matching_keys.extend(list(keys))
+
+        await client.quit()
+
+        return [self.parse_key(key.decode()) for key in matching_keys]
+
+    def get_in_memory_limit_keys(
+        self,
+    ) -> list[tuple[RateLimitItem, tuple[str, str | MatchAllInputs]]]:
+        """Returns a list of current RateLimitItems with tuples of namespace and primary key."""
+        return [self.parse_key(key) for key in self.storage.events]
+
+    async def get_limit_keys(
+        self,
+    ) -> list[tuple[RateLimitItem, tuple[str, str | MatchAllInputs]]]:
+        if os.environ.get("REDIS_URL") and not self.use_in_memory:
+            return await self.get_rate_limit_keys()
+        return self.get_in_memory_limit_keys()
+
+    async def rate_limit_status(self):
+
+        limit_status = {}
+
+        for rate_limit, (namespace, primary_key) in await self.get_limit_keys():
+            period_start, n_items_in_period = await self.storage.get_moving_window(
+                rate_limit.key_for(*(namespace, primary_key or "")),
+                rate_limit.amount,
+                rate_limit.get_expiry(),
+            )
+            limit_status[(namespace, primary_key)] = {
+                "period_start": period_start,
+                "n_items_in_period": n_items_in_period,
+                "period_seconds": rate_limit.GRANULARITY.seconds,
+                "period_name": rate_limit.GRANULARITY.name,
+                "period_cap": rate_limit.amount,
+            }
+        return limit_status
+
+    async def try_acquire(
+        self,
+        namespace_and_key: tuple[str, str | MatchAllInputs],
+        rate_limit: RateLimitItem | str | None = None,
+        machine_id: int = 0,
+        timeout: float = GLOBAL_RATE_LIMITER_TIMEOUT,  # noqa: ASYNC109
```

Comment:
```suggestion
        acquire_timeout: float = GLOBAL_RATE_LIMITER_TIMEOUT,
```

https://docs.astral.sh/ruff/rules/async-function-with-timeout/#details

It's better to be more specific in the `timeout`'s name, and avoids `ASYNC109`